charger/openevse.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	options := []openevse.ClientOption{openevse.WithHTTPClient(c.helper.Client)}
+
+	if user != "" && password != "" {
+		basicAuthProvider, err := securityprovider.NewSecurityProviderBasicAuth(user, password)
+		if err != nil {
+			return c, err
+		}
+
+		options = append(options, openevse.WithRequestEditorFn(basicAuthProvider.Intercept))
+	}
+
+	var err error
+	c.api, err = openevse.NewClientWithResponses(uri, options...)
+	if err != nil {
+		return c, err
+	}
+
+	var phases1p3p func(int) error
+	if err := c.hasPhaseSwitchCapabilities(); err == nil {
+		phases1p3p = c.phases1p3p
+
+		// disable EVSE's own 1/3-phase auto-switching
+		if err := c.rapiCommand("$S8 0"); err != nil {
+			return c, err
+		}
+	}
+
+	return decorateOpenEVSE(c, phases1p3p), err
+}
+
+func (c *OpenEVSE) requestContextWithTimeout() (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	return ctx, cancel
+}
+
+func (c *OpenEVSE) hasPhaseSwitchCapabilities() error {
+	return c.rapiCommand("$G7")
+}
+
+func (c *OpenEVSE) rapiCommand(command string) error {
+	var res struct {
+		Cmd, Ret string
+	}
+
+	uri := fmt.Sprintf("%s/r?json=1&rapi=%s", c.uri, url.QueryEscape(command))
+
+	err := c.helper.GetJSON(uri, &res)
+	if err == nil && !strings.HasPrefix(res.Ret, "$OK") {
+		err = fmt.Errorf("rapi command failed: %s", res.Ret)
+	}
+
+	return err
+}
+
+func (c *OpenEVSE) SetManualOverride(enable bool) error {
+	state := "disabled"
+	if enable {
+		state = "active"
+	}
+
+	data := openevse.SetManualOverrideJSONRequestBody{
+		State: &state,
+	}
+
+	ctx, cancel := c.requestContextWithTimeout()
+	defer cancel()
+	_, err := c.api.SetManualOverrideWithResponse(ctx, data)
+
+	return err
+}
+
+func (c *OpenEVSE) Status() (api.ChargeStatus, error) {
+	ctx, cancel := c.requestContextWithTimeout()
+	defer cancel()
+
+	res, err := c.api.GetStatusWithResponse(ctx)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	/*
+		0: "unknown",
+		1: "not connected",
+		2: "connected",
+		3: "charging",
+		4: "vent required",
+		5: "diode check failed",
+		6: "gfci fault",
+		7: "no ground",
+		8: "stuck relay",
+		9: "gfci self-test failure",
+		10: "over temperature",
+		11: "over current",
+		254: "sleeping",
+		255: "disabled"
+	*/
+
+	switch state := *res.JSON200.State; state {
+	case 1:
+		return api.StatusA, nil
+	case 2, 254, 255:
+		if connected := *res.JSON200.Vehicle != 0; connected {
+			return api.StatusB, nil
+		}
+		return api.StatusA, nil
+	case 3:
+		return api.StatusC, nil
+	case 4:
+		return api.StatusD, nil
+	case 5, 6, 7, 8, 9, 10, 11:
+		return api.StatusF, nil
+	default:
+		return api.StatusNone, fmt.Errorf("unknown EVSE state: %d", state)
+	}
+}
+
+// Enabled implements the api.Charger interface
+// Since OpenEVSE has many possible states, we manually override the state so it cannot change automatically
+// For example, there can be an energy limit set, or a timer to only charge between certain hours
```

Comment:
I don't like this. We could check the state and return an error if its not on manual. We should however NOT continuously work against another controlling mechanism. Either use evcc or something else. The entire complex logic below should rather be removed.